### PR TITLE
[Ray Log] remove glog dependency

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

We use spdlog for logging now, however we still rely on glog, because we need its GetCallTrace method, actually it can be replaced with absl.stacktrace, so this pr will remove glog dependency.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

**Comment by kfstorm** [comment]
@qicosmos Can you try to send SIGABRT to raylet and paste the stack trace logs with and without glog here?

**Comment by kfstorm** [comment]
We should delete this and all glog-related patches.

https://github.com/ray-project/ray/blob/498da139447e953c080d0ba9f02d6f0868a32a44/bazel/ray_deps_setup.bzl#L184-L194

File: src/ray/util/logging.cc [hunk 1]
```diff
@@ -64,36 +54,23 @@ std::string RayLog::log_format_pattern_ = "[%Y-%m-%d %H:%M:%S,%e %L %P %t] %v";
...
-
-  virtual void Write(bool /* should flush */, time_t /* timestamp */, const char *message,
-                     int length) {
-    // note: always flush otherwise it never shows up in raylet.out
-    out().write(message, length) << std::flush;
+  std::vector<void *> local_stack;
+  local_stack.resize(50);
+  absl::GetStackTrace(local_stack.data(), 50, 1);
+  static constexpr size_t buf_size = 1 << 14;
+  std::unique_ptr<char[]> buf(new char[buf_size]);
```
Comment (kfstorm) [inline src/ray/util/logging.cc:63]: Why do we need unique_ptr here?

File: src/ray/util/logging.cc [hunk 2]
```diff
@@ -64,36 +54,23 @@ std::string RayLog::log_format_pattern_ = "[%Y-%m-%d %H:%M:%S,%e %L %P %t] %v";
...
-  std::ostream &out() { return std::cout; }
-
-  virtual void Write(bool /* should flush */, time_t /* timestamp */, const char *message,
-                     int length) {
-    // note: always flush otherwise it never shows up in raylet.out
-    out().write(message, length) << std::flush;
+  std::vector<void *> local_stack;
+  local_stack.resize(50);
+  absl::GetStackTrace(local_stack.data(), 50, 1);
+  static constexpr size_t buf_size = 1 << 14;
```
Comment (kfstorm) [inline src/ray/util/logging.cc:62]:
```suggestion
  static constexpr size_t buf_size = 16 * 1024;
```

File: src/ray/util/logging.cc [hunk 3]
```diff
@@ -291,11 +241,7 @@ void RayLog::StartRayLog(const std::string &app_name, RayLogLevel severity_thres
   severity_threshold_ = severity_threshold;
   app_name_ = app_name;
   log_dir_ = log_dir;
-#if defined(RAY_USE_GLOG) || defined(RAY_USE_SPDLOG)
-#ifdef RAY_USE_GLOG
-  google::InitGoogleLogging(app_name_.c_str());
-  int level = GetMappedSeverity(static_cast<RayLogLevel>(severity_threshold_));
-#endif
+#ifdef RAY_USE_SPDLOG
```
Comment (kfstorm) [inline src/ray/util/logging.cc:244]: Can we remove `RAY_USE_SPDLOG` as well?

File: src/ray/util/logging.cc [hunk 4]
```diff
@@ -450,25 +349,13 @@ void WriteFailureMessage(const char *data, int size) {
...
   SetErrorMode(GetErrorMode() | SEM_FAILCRITICALERRORS);
   // If process crashes, don't display an error window.
   SetErrorMode(GetErrorMode() | SEM_NOGPFAULTERRORBOX);
 #endif
-#if defined(RAY_USE_GLOG) || defined(RAY_USE_SPDLOG)
-  if (is_failure_signal_handler_installed_) {
-    return;
-  }
-  google::InstallFailureSignalHandler();
-  google::InstallFailureWriter(&WriteFailureMessage);
```
Comment (kfstorm) [inline src/ray/util/logging.cc:469]: If we remove this, who will call `WriteFailureMessage`? I see that you removed all the signal handler-related code, but if we don't install the signal handler, how would we print stack trace on receiving a signal?

File: src/ray/util/logging.h [hunk 5]
```diff
@@ -43,10 +43,7 @@ enum { ERROR = 0 };
 #endif
 
 namespace ray {
-/// In order to use the get stacktrace method in other non-glog scenarios, we
-/// have added a patch to allow glog to return the current call stack information
-/// through the internal interface. This function `GetCallTrace` is a wrapper
-/// providing a new detection function for debug or something like that.
+/// This function return the current call stack information.
```
Comment (kfstorm) [inline src/ray/util/logging.h:46]:
```suggestion
/// This function returns the current call stack information.
```

File: src/ray/util/logging.cc (hunk 4 above)
Comment (qicosmos) [reply src/ray/util/logging.cc:469]: Maybe it should not be removed, i will test signal handler.

**Comment by qicosmos** [comment]
> @qicosmos Can you try to send SIGABRT to raylet and paste the stack trace logs with and without glog here?

I will do a test.

File: src/ray/util/logging.cc (hunk 4 above)
Comment (ashione) [reply src/ray/util/logging.cc:469]: We need to rewrite singalhandler if these functions of glog should be removed.

Comment (qicosmos) [reply src/ray/util/logging.cc:469]: Right, i will deal with signal handler later.

**Comment by rkooo567** [comment]
Can you also post some of stacktrace example? (e.g., show me how stacktrace looks like when SIGBUS happens)

**Comment by rkooo567** [comment]
@qicosmos can you post how the stacktrace looks like? Also does it make sense to write a python test to see if it prints logs on some signals? (especially for SIGSEGV, SIGBART, and SIGBUS)

**Comment by qicosmos** [comment]
> @qicosmos can you post how the stacktrace looks like? Also does it make sense to write a python test to see if it prints logs on some signals? (especially for SIGSEGV, SIGBART, and SIGBUS)

signal_test will print the stacktrace, like this:

`
[ RUN      ] SignalTest.SendBusSignalTest
*** SIGBUS received at time=1622527644 ***
[2021-06-01 14:07:24,263 E 99852 99852] signal_test.cc:42: SendBusSignalTest: kill pid 99866 with return value=0
[2021-06-01 14:07:24,263 E 99852 99852] signal_test.cc:42: SendBusSignalTest: kill pid 99866 with return value=0
PC: @     0x5608e45461b2  (unknown)  ray::TestSendSignal()
    @     0x7f624ca6e85d         64  absl::lts_2019_08_08::WriteFailureInfo()
    @     0x7f624ca6e9ff         96  absl::lts_2019_08_08::AbslFailureSignalHandler()
    @     0x7f623b73a5d0  246824784  (unknown)
    @     0x5608e45463fe        112  ray::SignalTest_SendBusSignalTest_Test::TestBody()
    @     0x7f623ba16475         48  testing::internal::HandleSehExceptionsInMethodIfSupported<>()
    @     0x7f623ba12127        144  testing::internal::HandleExceptionsInMethodIfSupported<>()
    @     0x7f623b9fe678        112  testing::Test::Run()
    @     0x7f623b9fefb0        112  testing::TestInfo::Run()
    @     0x7f623b9ff5fd        112  testing::TestCase::Run()
    @     0x7f623ba09783        112  testing::internal::UnitTestImpl::RunAllTests()
    @     0x7f623ba172fb         48  testing::internal::HandleSehExceptionsInMethodIfSupported<>()
    @     0x7f623ba12e85        144  testing::internal::HandleExceptionsInMethodIfSupported<>()
    @     0x7f623ba082ba        128  testing::UnitTest::Run()
    @     0x5608e45472b8         16  RUN_ALL_TESTS()
    @     0x5608e4546bbd         64  main
    @     0x7f623af64401  (unknown)  __libc_start_main
    @ 0x412e258d4c544155  (unknown)  (unknown)
`

File: src/ray/util/logging.cc (hunk 3 above)
Comment (qicosmos) [reply src/ray/util/logging.cc:246]: `
#ifdef RAY_USE_SPDLOG
typedef ray::SpdLogMessage LoggingProvider;
#else
typedef ray::CerrLog LoggingProvider;
#endif
`
If we don't define RAY_USE_SPDLOG, ray::CerrLog will be used. If we remove RAY_USE_SPDLOG, how to provide a default logging provider?

Comment (kfstorm) [reply src/ray/util/logging.cc:246]: I don't think `CerrLog` is used anymore. We can delete it as well.

**Comment by kfstorm** [comment]
@qicosmos Could you also provide the output if we use glog? One thing I'm concerned about is that the thread ID is missing in your pasted output, so we cannot identify which thread receives/sends the signal. Thread ID is quite useful when we encounter SIGSEGV.

BTW, there's no assertion logic in `signal_test.cc`. We have to verify the stack trace output manually. Also, the test doesn't write logs to log files, which is what we also care for.

@qicosmos Could you run a simple Ray application and then kill the Raylet and take the stack trace logs from the Raylet log files?

File: bazel/ray_deps_setup.bzl [hunk 6]
```diff
@@ -13,12 +13,13 @@ def urlsplit(url):
         split_on_scheme = [None] + split_on_scheme[:1]
     split_on_path = split_on_scheme[1].split("/")
     return {
-        "scheme": split_on_scheme[0],
-        "netloc": split_on_path[0].split("."),
-        "path": split_on_path[1:],
-        "query": split_on_query[1].split("&") if len(split_on_query) > 1 else None,
-        "fragment": split_on_anchor[1] if len(split_on_anchor) > 1 else None,
-    }
+  "scheme" : split_on_scheme[0],
```
Comment (kfstorm) [inline bazel/ray_deps_setup.bzl:16]: Please revert unrelated changes in this file

File: src/ray/util/logging_test.cc [hunk 7]
```diff
@@ -116,10 +116,13 @@ std::string TestFunctionLevel2() {
 #ifndef _WIN32
 TEST(PrintLogTest, CallstackTraceTest) {
   auto ret0 = TestFunctionLevel0();
+  std::cout << ret0 << std::endl;
```
Comment (kfstorm) [inline src/ray/util/logging_test.cc:119]: Revert this file.

File: src/ray/util/logging.cc [hunk 8]
```diff
@@ -67,33 +58,21 @@ long RayLog::log_rotation_file_num_ = 10;
...
-    out().write(message, length) << std::flush;
+  std::vector<void *> local_stack;
+  local_stack.resize(50);
+  absl::GetStackTrace(local_stack.data(), 50, 0);
+  static constexpr size_t buf_size = 16 * 1024;
+  char buf[buf_size];
+  std::string output;
+  for (auto &stack : local_stack) {
+    if (absl::Symbolize(stack, buf, buf_size)) {
+      output.append("    ").append(buf).append("\n");
```
Comment (rkooo567) [inline src/ray/util/logging.cc:69]: Hey is there any good way to add line information using addr2line here? https://linux.die.net/man/1/addr2line

Comment (rkooo567) [reply src/ray/util/logging.cc:69]: I am just curious if that's possible since it looks like we can access each stack trace

File: src/ray/util/logging.cc (hunk 3 above)
Comment (rkooo567) [reply src/ray/util/logging.cc:246]: I am okay with just deleting glog to make iteration faster btw. (and we can remove Cerrlog in the separate PR)

File: src/ray/util/logging_test.cc [hunk 9]
```diff
@@ -116,10 +116,13 @@ std::string TestFunctionLevel2() {
...
   EXPECT_TRUE(ret0.find("TestFunctionLevel0") != std::string::npos);
   auto ret1 = TestFunctionLevel1();
+  std::cout << ret1 << std::endl;
   EXPECT_TRUE(ret1.find("TestFunctionLevel1") != std::string::npos);
   auto ret2 = TestFunctionLevel2();
+  std::cout << ret2 << std::endl;
   EXPECT_TRUE(ret2.find("TestFunctionLevel2") != std::string::npos);
 }
```
Comment (rkooo567) [inline src/ray/util/logging_test.cc:127]: Can you add a python test? Maybe doing like 

Start a raylet -> send a signal -> check log file -> make sure stacktrace is in the .err file

**Review by rkooo567 — COMMENTED** [review]
A couple last comments!

**Comment by qicosmos** [comment]
> @qicosmos Could you also provide the output if we use glog? One thing I'm concerned about is that the thread ID is missing in your pasted output, so we cannot identify which thread receives/sends the signal. Thread ID is quite useful when we encounter SIGSEGV.
> 
> BTW, there's no assertion logic in `signal_test.cc`. We have to verify the stack trace output manually. Also, the test doesn't write logs to log files, which is what we also care for.
> 
> @qicosmos Could you run a simple Ray application and then kill the Raylet and take the stack trace logs from the Raylet log files?

This PR logging out:
```
[2021-06-21 16:13:38,663 E 15234 15234] signal_test.cc:42: SendBusSignalTest: kill pid 15236 with return value=0
[2021-06-21 16:13:38,663 E 15234 15234] signal_test.cc:42: SendBusSignalTest: kill pid 15236 with return value=0
PC: @     0x55eedf9a88c0  (unknown)  ray::TestSendSignal()
    @     0x7fc50d3704ed         80  absl::lts_2019_08_08::AbslFailureSignalHandler()
    @     0x7fc50c1a15d0  (unknown)  (unknown)
    @     0x55eedf9a8ad9         80  ray::SignalTest_SendBusSignalTest_Test::TestBody()
    @     0x7fc50c4113b7         96  testing::internal::HandleExceptionsInMethodIfSupported<>()
    @     0x7fc50c4115f2         48  testing::Test::Run()
    @     0x7fc50c4118e8         80  testing::TestInfo::Run()
    @     0x7fc50c411b75         80  testing::TestCase::Run()
    @     0x7fc50c411e18        144  testing::internal::UnitTestImpl::RunAllTests()
    @     0x7fc50c412127         96  testing::internal::HandleExceptionsInMethodIfSupported<>()
    @     0x7fc50c412322        128  testing::UnitTest::Run()
    @     0x55eedf9a7fd2        144  main
    @     0x7fc50b9c9401  (unknown)  __libc_start_main
    @ 0x44fe258d4c544155  (unknown)  (unknown)
[2021-06-21 16:13:38,667 E 15236 15234] logging.cc:299: *** SIGBUS received at time=1624263218 ***
[2021-06-21 16:13:38,667 E 15236 15234] logging.cc:299: *** SIGBUS received at time=1624263218 ***
[2021-06-21 16:13:38,667 E 15236 15234] logging.cc:299: PC: @     0x55eedf9a88c0  (unknown)  ray::TestSendSignal()
[2021-06-21 16:13:38,667 E 15236 15234] logging.cc:299: PC: @     0x55eedf9a88c0  (unknown)  ray::TestSendSignal()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50d370516         80  absl::lts_2019_08_08::AbslFailureSignalHandler()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50d370516         80  absl::lts_2019_08_08::AbslFailureSignalHandler()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c1a15d0  (unknown)  (unknown)
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c1a15d0  (unknown)  (unknown)
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x55eedf9a8ad9         80  ray::SignalTest_SendBusSignalTest_Test::TestBody()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x55eedf9a8ad9         80  ray::SignalTest_SendBusSignalTest_Test::TestBody()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c4113b7         96  testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c4113b7         96  testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c4115f2         48  testing::Test::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c4115f2         48  testing::Test::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c4118e8         80  testing::TestInfo::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c4118e8         80  testing::TestInfo::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c411b75         80  testing::TestCase::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c411b75         80  testing::TestCase::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c411e18        144  testing::internal::UnitTestImpl::RunAllTests()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c411e18        144  testing::internal::UnitTestImpl::RunAllTests()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c412127         96  testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c412127         96  testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c412322        128  testing::UnitTest::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50c412322        128  testing::UnitTest::Run()
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x55eedf9a7fd2        144  main
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x55eedf9a7fd2        144  main
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50b9c9401  (unknown)  __libc_start_main
[2021-06-21 16:13:38,668 E 15236 15234] logging.cc:299:     @     0x7fc50b9c9401  (unknown)  __libc_start_main
[2021-06-21 16:13:38,670 E 15236 15234] logging.cc:299:     @ 0x44fe258d4c544155  (unknown)  (unknown)
[2021-06-21 16:13:38,670 E 15236 15234] logging.cc:299:     @ 0x44fe258d4c544155  (unknown)  (unknown)
[       OK ] SignalTest.SendBusSignalTest (201 ms)
[ RUN      ] SignalTest.SIGABRT_Test
```

The glog logging out:
```
[2021-06-10 11:01:45,423 E 94936 94936] signal_test.cc:42: SendBusSignalTest: kill pid 94953 with return value=0
[2021-06-10 11:01:45,423 E 94936 94936] signal_test.cc:42: SendBusSignalTest: kill pid 94953 with return value=0
[2021-06-10 11:01:45,423 E 94953 94936] logging.cc:440: *** Aborted at 1623294105 (unix time) try "date -d @1623294105" if you are using GNU date ***
[2021-06-10 11:01:45,423 E 94953 94936] logging.cc:440: *** Aborted at 1623294105 (unix time) try "date -d @1623294105" if you are using GNU date ***
[2021-06-10 11:01:45,426 E 94953 94936] logging.cc:440: PC: @                0x0 (unknown)
[2021-06-10 11:01:45,426 E 94953 94936] logging.cc:440: PC: @                0x0 (unknown)
[2021-06-10 11:01:45,426 E 94953 94936] logging.cc:440: *** SIGBUS (@0x1f4000172d8) received by PID 94953 (TID 0x7f685213f940) from PID 94936; stack trace: ***
[2021-06-10 11:01:45,426 E 94953 94936] logging.cc:440: *** SIGBUS (@0x1f4000172d8) received by PID 94953 (TID 0x7f685213f940) from PID 94936; stack trace: ***
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f6838efc5d0 (unknown)
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f6838efc5d0 (unknown)
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x561b23e24710 ray::TestSendSignal()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x561b23e24710 ray::TestSendSignal()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x561b23e24929 ray::SignalTest_SendBusSignalTest_Test::TestBody()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x561b23e24929 ray::SignalTest_SendBusSignalTest_Test::TestBody()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916ab67 testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916ab67 testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916ada2 testing::Test::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916ada2 testing::Test::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b098 testing::TestInfo::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b098 testing::TestInfo::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b325 testing::TestCase::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b325 testing::TestCase::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b5c8 testing::internal::UnitTestImpl::RunAllTests()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b5c8 testing::internal::UnitTestImpl::RunAllTests()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b8d7 testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916b8d7 testing::internal::HandleExceptionsInMethodIfSupported<>()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916bad2 testing::UnitTest::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x7f683916bad2 testing::UnitTest::Run()
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x561b23e23e52 main
[2021-06-10 11:01:45,427 E 94953 94936] logging.cc:440:     @     0x561b23e23e52 main
[2021-06-10 11:01:45,428 E 94953 94936] logging.cc:440:     @     0x7f6838726401 __libc_start_main
[2021-06-10 11:01:45,428 E 94953 94936] logging.cc:440:     @     0x7f6838726401 __libc_start_main
[2021-06-10 11:01:45,428 E 94953 94936] logging.cc:440:     @     0x561b23e23efa _start
[2021-06-10 11:01:45,428 E 94953 94936] logging.cc:440:     @     0x561b23e23efa _start
[       OK ] SignalTest.SendBusSignalTest (201 ms)
[ RUN      ] SignalTest.SIGABRT_Test
```

**Comment by kfstorm** [comment]
@qicosmos According to the outputs you posted above, it seems that there's one more repeat of stack trace when glog is removed. Do you know why?

**Comment by qicosmos** [comment]
> @qicosmos According to the outputs you posted above, it seems that there's one more repeat of stack trace when glog is removed. Do you know why?

Because absl will output the stacks to stderr, [here is the source code](https://github.com/abseil/abseil-cpp/blob/master/absl/debugging/failure_signal_handler.cc#L364), it's no side effects.

**Comment by kfstorm** [comment]
> > @qicosmos According to the outputs you posted above, it seems that there's one more repeat of stack trace when glog is removed. Do you know why?
> 
> Because absl will output the stacks to stderr firstly, [here is the source code](https://github.com/abseil/abseil-cpp/blob/master/absl/debugging/failure_signal_handler.cc?rgh-link-date=2021-06-21T13%3A28%3A24Z#L364), it's no side effects.

OK, it seems that this behavior cannot be turned off.

File: python/ray/tests/test_kill_raylet_signal_log.py [hunk 10]
```diff
@@ -1,28 +1,51 @@
...
-    with open(raylet_out_path) as f:
-        s = f.read()
-        assert "SIGABRT" in s
+    raylet_out_path = "{}/logs/raylet.out".format(session_dir)
+    assert len(raylet_out_path) > 0
+    pid = ray.get(f.remote())
+    os.kill(pid, SIGKILL)
+    wait_for_pid_to_exit(pid)
+    thread = check_thread(raylet_out_path)
+    thread.start()
```
Comment (kfstorm) [inline python/ray/tests/test_kill_raylet_signal_log.py:48]: Forgot to join the thread.
Q1: Why do we need a thread here?
Q2: Will the exception thrown in the thread be populated to the main thread?

File: python/ray/tests/test_kill_raylet_signal_log.py [hunk 11]
```diff
@@ -1,28 +1,51 @@
...
-def get_pid(name):
-    pids = psutil.process_iter()
-    for pid in pids:
-        if (pid.name() == name):
-            return pid.pid
+
+class check_thread(threading.Thread):
+    def __init__(self, raylet_out_path):
+        threading.Thread.__init__(self)
+        self._daemonic = True
```
Comment (kfstorm) [inline python/ray/tests/test_kill_raylet_signal_log.py:28]: Why set this?

File: python/ray/tests/test_kill_raylet_signal_log.py [hunk 12]
```diff
@@ -1,28 +1,51 @@
...
+SIGKILL = signal.SIGKILL if sys.platform != "win32" else signal.SIGTERM
+
+
+def check_result(path):
+    with open(path) as f:
+        time.sleep(1)
+        s = f.read()
+        assert len(s) > 0
+        print(s)
+        assert "SIGTERM" in s
```
Comment (kfstorm) [inline python/ray/tests/test_kill_raylet_signal_log.py:22]: For non-Windows environments, you send a SIGKILL to Raylet. Why do we always assert SIGTERM here?

File: python/ray/tests/test_kill_raylet_signal_log.py (hunk 11 above)
Comment (qicosmos) [reply python/ray/tests/test_kill_raylet_signal_log.py:28]: The _daemonic make the thread detach.
If we use thread.join() here, the log of SIGTERM can not be flushed, use detach thread can avoid the problem.

File: python/ray/tests/test_kill_raylet_signal_log.py (hunk 10 above)
Comment (qicosmos) [reply python/ray/tests/test_kill_raylet_signal_log.py:48]: FYI: https://github.com/ray-project/ray/pull/16077#discussion_r660738213

File: python/ray/tests/test_kill_raylet_signal_log.py (hunk 12 above)
Comment (qicosmos) [reply python/ray/tests/test_kill_raylet_signal_log.py:22]: Windows no SIGABRT, but has SIGTERM, use SIGTERM to support windows and linux.

File: python/ray/tests/test_kill_raylet_signal_log.py (hunk 11 above)
Comment (kfstorm) [reply python/ray/tests/test_kill_raylet_signal_log.py:28]: But in this case, the test will pass without waiting for the thread to finish. How do you ensure the signal is in the log file?

File: python/ray/tests/test_kill_raylet_signal_log.py (hunk 10 above)
Comment (kfstorm) [reply python/ray/tests/test_kill_raylet_signal_log.py:48]: You still didn't answer Q1 and Q2.

File: python/ray/tests/test_kill_raylet_signal_log.py (hunk 12 above)
Comment (kfstorm) [reply python/ray/tests/test_kill_raylet_signal_log.py:22]: For Linux and macOS, you send SIGKILL according to `signal.SIGKILL if sys.platform != "win32" else signal.SIGTERM`, right? But you assert `SIGTERM` here. So I suppose the test will fail on Linux and macOS.

**Review by kfstorm — APPROVED** [review]

**Comment by kfstorm** [comment]
@rkooo567 Any more comments?

**Review by rkooo567 — APPROVED** [review]
